dynamicio/io/serde.py

```diff
@@ -0,0 +1,114 @@
+"""These are the base serde classes, used for testing & when appropriate for actual IO."""
+import pickle
+from abc import ABC, abstractmethod
+from io import BytesIO
+from threading import Lock
+from typing import Callable, Optional, TypeVar, Union
+
+import pandas as pd
+from uhura.serde import Serde
+
+from dynamicio import utils
+from dynamicio.io.hdf import HdfIO
+
+SerdeType = TypeVar("SerdeType")
+
+
+class BaseSerde(ABC, Serde[pd.DataFrame]):
+    file_extension = "_"
+
+    def __init__(self, validations: Optional[Callable] = None, **kwargs):
+        self.validations = validations or []
+
+    def read_from_file(self, file) -> pd.DataFrame:
+        df = self._read(file)
+        return self.validate(df)
+
+    @abstractmethod
+    def _read(self, file) -> pd.DataFrame:
+        raise NotImplementedError
+
+    def write_to_file(self, path: str, obj: pd.DataFrame) -> None:
+        return self._write(path, obj)
+
+    @abstractmethod
+    def _write(self, path: str, obj: pd.DataFrame) -> None:
+        raise NotImplementedError
+
+    def validate(self, df: pd.DataFrame):
+        """Validation is done here to avoid double validations in the framework."""
+        for validator in self.validations:
+            validator(df)
+        return df
+
+
+class PickleSerde(BaseSerde):
+    def _read(self, file) -> SerdeType:
```

Comment:
what is `SerdeType`?

Reply:
`SerdeType = TypeVar("SerdeType")` 😁

Basically any object that might be written or read? I.e. pickled or unpickled I think.

Reply:
I think it's just taken as an example from here and could probably deleted in dynamicio. https://github.com/VorTECHsa/uhura/blob/master/uhura/serde.py